{
 "metadata": {
  "name": "",
  "signature": "sha256:b4122879241504017a066a3a6c4f776f55904fb3081864349f32323f8d164a54"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#Guide to Running \n",
      "\n",
      "##Initialization\n",
      "* [__download_data__](./download_data.ipynb)  \n",
      "  Pulls all of the necessary data from the net and constructs the file tree and data objects used in the rest of the analysis. \n",
      "  \n",
      "  \n",
      "* [__get_all_MAFs__](./get_all_MAFs.ipynb)  \n",
      "  Script to download and process updated MAF files from the TCGA Data Portal. \n",
      "  \n",
      "  \n",
      "* [__get_updated_clinical__](./get_updated_clinical.ipynb)  \n",
      "  Script to download and process updated clinical data from the TCGA Data Portal.\n",
      "  \n",
      "  \n",
      "##Primary Analysis  \n",
      "(There are dependencies among these, run them in order.)\n",
      "* [__HPV_Process_Data__](./HPV_Process_Data.ipynb)  \n",
      "  Compile HPV status for all patient tumors.  \n",
      "  Calculate global variables and meta features in the HPV- background. \n",
      "  \n",
      "* [__binarize_clinical__](./binarize_clinical.ipynb)  \n",
      "  Process clinical variables into binary matrix for use in prognostic screens.   \n",
      "\n",
      "* [__Prognostic_Screen__](./Prognostic_Screen.ipynb)  \n",
      "  Run the primary prognostic screen for HPV- HNSCC patients.  \n",
      "  \n",
      "  \n",
      "* [__Secondary_Screen__](./Secondary_Screen.ipynb)   \n",
      "  Run the prognostic screen for HPV- HNSCC patients with the TP53-3p event.\n",
      "  \n",
      "  \n",
      "* [__HNSCC_figures__](./HNSCC_figures.ipynb)  \n",
      "  Generate some of the figure panels for the HNSCC discovery cohort.  Some of the other figures and figure panels are generated inline with analysis. \n",
      "  \n",
      "  \n",
      "##Validation Cohorts\n",
      "\n",
      "* [__UPMC_cohort__](./UPMC_cohort.ipynb)  \n",
      "  Validation of primary findings in independent patient cohort from University of Pittsburgh ([Stansky et al.](http://www.sciencemag.org/content/333/6046/1157.full)).\n",
      "  \n",
      "\n",
      "* [__Molecular_Validation__](./Molecular_Validation.ipynb)  \n",
      "Validation of molecular associations in recent TCGA samples.\n",
      "\n",
      "\n",
      "* [__PANCAN_cohort__](./PANCAN_cohort.ipynb)  \n",
      "  Validation of primary findings across ~4400 TCGA patient tumors.  \n",
      "  \n",
      "\n",
      "\n",
      "\n",
      "  \n",
      "##Targeted Analysis for Support of Main Findings  \n",
      "\n",
      "* [__Reviewer_Response__](./Reviewer_Response.ipynb)  \n",
      "  Specific responses to reviewer comments.\n",
      "  \n",
      "\n",
      "* [__HNSCC_clinical_characterization__](HNSCC_clinical_characterization.ipynb)  \n",
      "  Overview of clinical variables in the TCGA HNSCC cohort and their implications towards patient prognosis.\n",
      "  \n",
      "\n",
      "* [__TP53_exploration__](./TP53_exploration.ipynb)  \n",
      "  Detailed characterization of TP53 mutations and their predicted functional impact. \n",
      "  \n",
      "  \n",
      "* [__HPV_characterization__](HPV_characterization.ipynb)  \n",
      "  Detailed characterization of the clinical and molecular coorelates of HPV+ status.  \n",
      "  \n",
      "\n",
      "* [__copy_number_exploration__](./copy_number_exploration.ipynb)  \n",
      "  Exploration of chromosomal instability, 3p deletion, TP53 mutation and the relationships between these factors.  \n",
      "  \n",
      "\n",
      "* [__Clinical_Covariates__](./Clinical_Covariates.ipynb)  \n",
      "  Exploration of primary subtypes within the context of a number of clinical variables. \n",
      "  \n",
      "  \n",
      "* [__Multivariate_Modeling__](./Multivariate_Modeling.ipynb)  \n",
      "  Exploration of primary subtypes within the context of a few different multivarite models including clinical variables.\n",
      "  \n",
      "  \n",
      "  \n",
      "##Variant Calling (optional)\n",
      "\n",
      "This requires a number of additional dependencies for sequencing analysis and as well as function calls to proprietary software installed on our virtual machine hosed by Annai Systems.  We have included all of the dependencies of this mutation calling step in the supplement as MAF files and highly recomend starting with these as opposed to recalling mutations. \n",
      "\n",
      "  \n",
      "* [__muTect_streamline__](muTect_streamline.ipynb)  \n",
      "  This script is used to generate bash scripts to download and process additional TCGA data from CGHub.  \n",
      "  \n",
      "  \n",
      "* [__new_data_process_TP53_Pancancer__](new_data_process_TP53_Pancancer.ipynb)  \n",
      "  Here we process the SNV and indel calls made by the variant calling tools, annotate them and consolidate them into a MAF file.\n",
      "  "
     ]
    }
   ],
   "metadata": {}
  }
 ]
}