{
 "metadata": {
  "name": "",
  "signature": "sha256:8fc9802f4d3ec1ab25199ef7a8ae59ef5d5088d37dbc5ea1eabb37caacc1237d"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#Software Overview \n",
      "\n",
      "This repository contains instructions for reproduction and extension of [A prognostic model of head and neck cancer ties TP53 mutation to 3p loss]() by Gross et al.\n",
      "\n",
      "##Dependencies \n",
      "\n",
      "This code uses a number of features in the scientific python stack as well as a small set of standard R libraries.  Thus far, this code has only been tested in a Linux enviroment, it may take some modification to run on other operating systems.\n",
      "\n",
      "I highly recomend installing a scientific Python distribution such as [Anaconda](http://continuum.io/) or [Enthought](https://www.enthought.com/) to handle the majority of the Python dependencies in this project (other than rPy2 and matplotlib_venn).  These are both free for academic use.\n",
      "\n",
      "###Python Dependencies  \n",
      "* [Numpy and Scipy](http://www.scipy.org/), numeric calculations and statistics in Python \n",
      "* [matplotlib](http://matplotlib.org/), plotting in Python\n",
      "* [Pandas](http://pandas.pydata.org/), data-frames for Python, handles the majority of data-structures  \n",
      "* [statsmodels](http://statsmodels.sourceforge.net/), used for statstics  \n",
      "* [scikit-learn](http://scikit-learn.org/stable/), used for supervised learning\n",
      "* [rPy2](http://rpy.sourceforge.net/rpy2.html), communication between R and Python  \n",
      "  * __NOT IN DISTRIBUTIONS__  \n",
      "  * I recommend installing with `pip install rpy2`  \n",
      "  * Needs R to be compiled with shared libraries  \n",
      "* [matplotlib_venn](https://pypi.python.org/pypi/matplotlib-venn) \n",
      "  * __NOT IN DISTRIBUTIONS__  \n",
      "  * I recommend installing with `pip install matplotlib_venn` \n",
      "  * Only used for Venn diagrams, not essential\n",
      "  \n",
      "  \n",
      "###R Dependencies\n",
      "* Needs to be compiled with shared libraries to communicate with Python (_this can be tricky_)\n",
      "* Packages\n",
      "  * base\n",
      "  * survival\n",
      "  * MASS\n",
      "   \n",
      "###Command Line Dependencies \n",
      "* curl (http://curl.haxx.se/) for fetching urls"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#Guide to Running \n",
      "\n",
      "##Initialization\n",
      "* __download_data__  \n",
      "  Pulls all of the necessary data from the net and constructs the file tree and data objects used in the rest of the analysis. \n",
      "  \n",
      "  \n",
      "* __get_all_MAFs__  \n",
      "  Script to download and process updated MAF files from the TCGA Data Portal. \n",
      "  \n",
      "  \n",
      "* __get_updated_clinical__  \n",
      "  Script to download and process updated clinical data from the TCGA Data Portal.\n",
      "  \n",
      "  \n",
      "##Primary Analysis  \n",
      "(There are dependencies among these, run them in order.)\n",
      "* [__HPV_Process_Data__](./HPV_Process_Data.ipynb)  \n",
      "  Compile HPV status for all patient tumors.  \n",
      "  Calculate global variables and meta features in the HPV- background. \n",
      "  \n",
      "  \n",
      "* [__Clinical_Inference__](./Clinical_Inference.ipynb)  \n",
      "  Infer missing values for smoking, drinking, perineural invasion, and extra-capsular spread clinical variables using rna expression data. \n",
      "  \n",
      "\n",
      "* [__HNSCC_clinical_characterization__](HNSCC_clinical_characterization.ipynb)  \n",
      "  Overview of clinical variables in the TCGA HNSCC cohort and their implications towards patient prognosis.\n",
      "  \n",
      "  \n",
      "* [__Prognostic_Screen__](../Prognostic_Screen.ipynb)  \n",
      "  Run the primary prognostic screen for HPV- HNSCC patients.  \n",
      "  \n",
      "  \n",
      "* __HNSCC_biomarkers_next__   \n",
      "  Run the prognostic screen for HPV- HNSCC patients with the TP53-3p event.\n",
      "  \n",
      "  \n",
      "* __HNSCC_figures__  \n",
      "  Generate some of the figure panels for the HNSCC discovery cohort.  Some of the other figures and figure panels are generated inline with analysis. \n",
      "  \n",
      "  \n",
      "##Validation Cohorts\n",
      "\n",
      "* __UPMC_cohort__  \n",
      "  Validation of primary findings in independent patient cohort from University of Pittsburgh ([Stansky et al.](http://www.sciencemag.org/content/333/6046/1157.full)).\n",
      "  \n",
      "\n",
      "* [__Molecular_Validation__](./Molecular_Validation.ipynb)  \n",
      "Validation of molecular associations in recent TCGA samples.\n",
      "\n",
      "\n",
      "* __PANCAN_cohort__  \n",
      "  Validation of primary findings across ~4400 TCGA patient tumors.  \n",
      "  \n",
      "\n",
      "\n",
      "\n",
      "  \n",
      "##Targeted Analysis for Support of Main Findings  \n",
      "\n",
      "\n",
      "\n",
      "* __TP53_exploration__  \n",
      "  Detailed characterization of TP53 mutations and their predicted functional impact. \n",
      "  \n",
      "  \n",
      "* __HPV_characterization__  \n",
      "  Detailed characterization of the clinical and molecular coorelates of HPV+ status.  \n",
      "  \n",
      "\n",
      "* __copy_number_exploration__  \n",
      "  Exploration of chromosomal instability, 3p deletion, TP53 mutation and the relationships between these factors.  \n",
      "  \n",
      "\n",
      "* __HNSCC_subtypes_clinical_covariates__  \n",
      "  Exploration of primary subtypes within the context of a number of clinical variables. \n",
      "  \n",
      " \n",
      "* __HNSCC_subtypes_molecular__ \n",
      "  Characterization of molecular coorelates of the primary subtypes.  \n",
      "  Generates most figures that describe the TCGA HNSCC cohort.\n",
      "  \n",
      "  \n",
      "##Variant Calling (optional)\n",
      "\n",
      "This requires a number of additional dependencies for sequencing analysis and as well as function calls to proprietary software installed on our virtual machine hosed by Annai Systems.  We have included all of the dependencies of this mutation calling step in the supplement as MAF files and highly recomend starting with these as opposed to recalling mutations. \n",
      "\n",
      "  \n",
      "* __muTect_streamline__  \n",
      "  This script is used to generate bash scripts to download and process additional TCGA data from CGHub.  \n",
      "  \n",
      "  \n",
      "* __new_data_process_HNSCC__  \n",
      "  Here we process the SNV and indel calls made by the variant calling tools, annotate them and consolidate them into a MAF file.\n",
      "  \n",
      "  \n",
      "* __new_data_process_TP53_Pancancer__  \n",
      "  Here we process the SNV and indel calls made by the variant calling tools, annotate them and consolidate them into a MAF file.\n",
      "  "
     ]
    }
   ],
   "metadata": {}
  }
 ]
}